homeassistant/components/modbus/validators.py

```diff
@@ -72,20 +67,79 @@
 # These keywords are only legal with some datatype: ...
 # As expressed in DEFAULT_STRUCT_FORMAT
 
+REGISTER_COUNT_ENTRY = namedtuple("REGISTER_COUNT_ENTRY", ["count_for_register_size"])
+# REGISTER_COUNT_ENTRY defines the record count for each data type for a given register size in bytes:
+#   2: register count when size is 2 bytes
+#   4: register count when size is 4 bytes
+#   8: register count when size is 8 bytes
+# -1 when register size is not valid for that data_type
+
 DEFAULT_STRUCT_FORMAT = {
-    DataType.INT8: ENTRY("b", 1, PARM_IS_LEGAL(False, False, False, False, False)),
-    DataType.UINT8: ENTRY("c", 1, PARM_IS_LEGAL(False, False, False, False, False)),
-    DataType.INT16: ENTRY("h", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.UINT16: ENTRY("H", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.FLOAT16: ENTRY("e", 1, PARM_IS_LEGAL(False, False, True, True, False)),
-    DataType.INT32: ENTRY("i", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.UINT32: ENTRY("I", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.FLOAT32: ENTRY("f", 2, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.INT64: ENTRY("q", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.UINT64: ENTRY("Q", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.FLOAT64: ENTRY("d", 4, PARM_IS_LEGAL(False, False, True, True, True)),
-    DataType.STRING: ENTRY("s", -1, PARM_IS_LEGAL(True, False, False, False, False)),
-    DataType.CUSTOM: ENTRY("?", 0, PARM_IS_LEGAL(True, True, False, False, False)),
+    DataType.INT8: ENTRY(
+        "b",
+        REGISTER_COUNT_ENTRY({"2": -1, "4": -1, "8": -1}),
```

Comment:
This is not a very elegant way, it is very hard to understand.

Why not make a multi-key as I suggested earlier.

Reply:
That would mean 3x the size of the table so tried to avoid it. Can you type here how you'd like a row of the table to look like.

Reply:
Yes it would triple the table, but allow e.g. INT16 to be marked as invalid 

My solution is most likely not the only one, but important is to avoid a lot of “if” in the code by having it tabled instead

Reply:
You might also be able to have a second table specific for size…just a thought.

Reply:
The main reason for my solution was to save ifs by indexing the register count with the number of register bytes. Would need more ifs using namedtuple. I'm happy to do any solution wether a second table or having 3x the rows 

If you can type your example table it will give me the idea of how you'd like to see things (ie what is readable and how it can be indexed). 

For example what does INT16 be marked as invalid means ( talking about 4/8 bytes per register cases), is having -1 count acceptable?

Reply:
Have added the 2 tables: one for validation of parameters the other with the counts as suggested